### PR TITLE
stdenv: add meta.repository (fixed)

### LINKED ISSUE #293838: Feature request: new meta attr for source code repository
## Problem

most packages use `meta.homepage` to link to the source code repository, but for projects with an actual homepage (such as luajit), finding the source code often requires clicking around various links to try to find the official repo.

for packages that use `fetchFromGitHub` or similar, you can usually piece the url back together with some effort, but for projects that use a source tarball (again, like luajit), you're out of luck.

## Proposal

a `meta.repository` field that points to an http-browsable source tree.

for packages without a separate homepage, you could just set the `meta.repository` field instead.

## Checklist

- [ ] checked [latest Nixpkgs manual] \([source][nixpkgs-source]) and [latest NixOS manual] \([source][nixos-source])
- [ ] checked [open documentation issues] for possible duplicates
- [ ] checked [open documentation pull requests] for possible solutions

[latest Nixpkgs manual]: https://nixos.org/manual/nixpkgs/unstable/
[latest NixOS manual]: https://nixos.org/manual/nixos/unstable/
[nixpkgs-source]: https://github.com/NixOS/nixpkgs/tree/master/doc
[nixos-source]: https://github.com/NixOS/nixpkgs/tree/master/nixos/doc/manual
[open documentation issues]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+label%3A%229.needs%3A+documentation%22
[open documentation pull requests]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+is%3Apr+label%3A%228.has%3A+documentation%22%2C%226.topic%3A+documentation%22

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by lolbinarycat**
## Description of changes
packages assume `src` won't be evaluated on unsupported platforms, so we match that assumption.

hopefully this doesn't break anything else.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lolbinarycat** [comment]
@AndersonTorres @a-n-n-a-l-e-e because of their role in the previous PR.

**Comment by lolbinarycat** [comment]
the fix to serapi seems to be getting deleted in the merge, unsure how to fix.

**Comment by lolbinarycat** [comment]
i think this should work

File: pkgs/stdenv/generic/check-meta.nix
Comment (AndersonTorres) [inline pkgs/stdenv/generic/check-meta.nix:481]: Since the user is somewhat free to set `meta.repository`, the use of `//` can bring unexpected outcomes.

https://nix.dev/guides/best-practices.html#updating-nested-attribute-sets

It's better to run this logic inside the block (below `name = . . .;` below),  somewhat like

```nix
repository = 
  if (complicatedConditions) then
   complicatedFunction
  else
    [ ];
```

Comment (lolbinarycat) [reply pkgs/stdenv/generic/check-meta.nix:481]: > Since the user is somewhat free to set meta.repository, the use of // can bring unexpected outcomes.

what unexpected outcomes?

manually setting `repository` will override the autodetected value, but that is intended behavior.

that link specifically warns of the fact that `//` performs a shallow update, but a shallow update is what is desired here.

> It's better to run this logic inside the block (below name = . . .; below), somewhat like

if you want the field to default to unset (to match the behavior of the other fields), that is impossible, since nix has no equivalent to javascript's `undefined`.

Comment (AndersonTorres) [reply pkgs/stdenv/generic/check-meta.nix:485]: > if you want the field to default to unset

no. This is only strange.

But the surrounding code is no less strange.

Comment (lolbinarycat) [reply pkgs/stdenv/generic/check-meta.nix:485]: if you have a problem with defaulting to unset, take it up with the original authors of `check-meta.nix` or open a github issue, it wasn't my choice.

**Comment by lolbinarycat** [comment]
something is throwing an error and i don't know what.  if anyone can tell me what package is failing to eval, or why ofborg-eval dosn't show any error logs, that would be greatly appreciated.

**Comment by lolbinarycat** [comment]
if this doesn't work, then i'll try fixing eval errors at the package end instead of the check-meta end.  this could actually be done by manually adding `repository` to failing packages' meta attributes

File: pkgs/stdenv/generic/check-meta.nix
Comment (AndersonTorres) [reply pkgs/stdenv/generic/check-meta.nix:485]: I DO NOT LIKE DEFAULT TO UNSET

Can you please stop trying to read what I did not say?

Comment (AndersonTorres) [reply pkgs/stdenv/generic/check-meta.nix:485]: **I never said a thing about** "default to unset".

Comment (lolbinarycat) [reply pkgs/stdenv/generic/check-meta.nix:485]: > > if you want the field to default to unset
> 
> no. This is only strange.
> 
> But the surrounding code is no less strange.

is this not saying "defaulting to unset is strange"?

if that's not what you are trying to say, please elaborate, i can't quite understand that sentence.

Comment (AndersonTorres) [reply pkgs/stdenv/generic/check-meta.nix:485]: "No" means "This is not what was I talking about".

"Strange" is, more precisely, about the fact the code looks like "I need to grab a piece of information, but that piece of information is hidden after a long and fragile matrioska of nested attributes that can break in unexpected ways".

It should not be so hard to just query `src.meta.homepage`.

On the other hand, the whole check-meta file is equally hard to understand. That's what I meant.

Comment (lolbinarycat) [reply pkgs/stdenv/generic/check-meta.nix:485]: most of check-meta actually has to do with reccomending remediations for how to fix common errors (such as enabling NIXPKGS_ALLOW_UNFREE)

the actual verification is relatively simple.

> It should not be so hard to just query src.meta.homepage

no, but unfortunately we are running into a significant downside of lazy evaluation here, and there's not a ton we can do.

Comment (lolbinarycat) [reply pkgs/stdenv/generic/check-meta.nix:485]: also a downside of not having implicit null

Comment (a-n-n-a-l-e-e) [reply pkgs/stdenv/generic/check-meta.nix:485]: eliminating the optionalAttrs also eliminates the eval performance hit. https://github.com/NixOS/nixpkgs/pull/300177#issuecomment-2029698758

Comment (lolbinarycat) [reply pkgs/stdenv/generic/check-meta.nix:485]: hmm... i guess we could just make it default to the empty list and be a bit different.  i believe nixos search treats unset as an empty list anyways.

Comment (lolbinarycat) [reply pkgs/stdenv/generic/check-meta.nix:485]: additionally, not using // means the conditions are only evaluated if meta.repository is not already set, which means any packages that have a src that doesn't eval can just set `repository` manually.

**Comment by a-n-n-a-l-e-e** [comment]
results from various permutations of the change. diffs below.

|          name          |   baseline    | prlist % | prlist2 % | prlist3 % | prlist4 % |
|------------------------|--------------:|---------:|----------:|----------:|----------:|
| cpuTime                | 128.5         | 0.72     | 0.22      | -0.66     | -0.35     |
| envs.bytes             | 6500151112.0  | 0.22     | 0.21      | 0.23      | 0.22      |
| envs.elements          | 341667635.0   | 0.2      | 0.19      | 0.23      | 0.21      |
| envs.number            | 235425627.0   | 0.24     | 0.22      | 0.22      | 0.22      |
| gc.heapSize            | 19532808192.0 | -0.09    | -0.09     | -0.09     | -0.09     |
| gc.totalBytes          | 39995440312.0 | 0.24     | 0.23      | 0.24      | 0.24      |
| list.bytes             | 762081056.0   | 0.12     | 0.08      | 0.08      | 0.08      |
| list.concats           | 17115912.0    | 0.0      | 0.0       | 0.0       | 0.0       |
| list.elements          | 95260132.0    | 0.12     | 0.08      | 0.08      | 0.08      |
| nrAvoided              | 272612023.0   | 0.19     | 0.15      | 0.17      | 0.14      |
| nrFunctionCalls        | 217154209.0   | 0.24     | 0.23      | 0.19      | 0.19      |
| nrLookups              | 109933236.0   | 0.42     | 0.39      | 0.39      | 0.39      |
| nrOpUpdateValuesCopied | 605984516.0   | 0.12     | 0.12      | 0.12      | 0.12      |
| nrOpUpdates            | 28428461.0    | 0.53     | 0.53      | 0.53      | 0.53      |
| nrPrimOpCalls          | 121479871.0   | 0.27     | 0.16      | 0.16      | 0.16      |
| nrThunks               | 334647453.0   | 0.34     | 0.31      | 0.35      | 0.35      |
| sets.bytes             | 13825998032.0 | 0.22     | 0.22      | 0.22      | 0.22      |
| sets.elements          | 810928496.0   | 0.19     | 0.19      | 0.19      | 0.19      |
| sets.number            | 53196381.0    | 0.63     | 0.63      | 0.63      | 0.63      |
| sizes.Attr             | 16.0          | 0.0      | 0.0       | 0.0       | 0.0       |
| sizes.Bindings         | 16.0          | 0.0      | 0.0       | 0.0       | 0.0       |
| sizes.Env              | 16.0          | 0.0      | 0.0       | 0.0       | 0.0       |
| sizes.Value            | 24.0          | 0.0      | 0.0       | 0.0       | 0.0       |
| symbols.bytes          | 2326781.0     | 0.0      | 0.0       | 0.0       | 0.0       |
| symbols.number         | 167519.0      | 0.0      | 0.0       | 0.0       | 0.0       |
| values.bytes           | 10257889848.0 | 0.32     | 0.29      | 0.33      | 0.33      |
| values.number          | 427412077.0   | 0.32     | 0.29      | 0.33      | 0.33      |

using 127c06803636f1e2c1de2e63bd767cf39f8d6c2c
<details><summary>prlist.diff</summary>

```diff
diff --git a/pkgs/stdenv/generic/check-meta.nix b/pkgs/stdenv/generic/check-meta.nix
index 1cd1ae6dd72e..c092063a03e8 100644
--- a/pkgs/stdenv/generic/check-meta.nix
+++ b/pkgs/stdenv/generic/check-meta.nix
@@ -10,7 +10,9 @@ let
     concatMapStrings
     concatMapStringsSep
     concatStrings
+    filter
     findFirst
+    head
     isDerivation
     length
     concatMap
@@ -303,6 +305,10 @@ let
       (listOf str)
       str
     ];
+    repository = union [
+      (listOf str)
+      str
+    ];
     downloadPage = str;
     changelog = union [
       (listOf str)
@@ -440,6 +446,22 @@ let
     # -----
     else { valid = "yes"; });
 
+  # get the default value for the meta.repository field.
+  # the fetchFrom* fetchers set src.meta.homepage
+  getRepository = let
+    getSrcs = attrs:
+      filter (src: src ? meta.homepage)
+        (if attrs ? src
+         then [ attrs.src ]
+         else optionals (attrs ? srcs && isList attrs.srcs) attrs.srcs);
+    getHomePages = map (src: src.meta.homepage);
+    unlist = list:
+      # the following line should show up in the stack trace is src doesn't eval
+      # HINT: try adding meta.repository to your package if it fails to eval
+      if length list == 1
+      then head list
+      else list;
+    in attrs: unlist (getHomePages (getSrcs attrs));
 
   # The meta attribute is passed in the resulting attribute set,
   # but it's not part of the actual derivation, i.e., it's not
@@ -452,8 +474,9 @@ let
     let
       outputs = attrs.outputs or [ "out" ];
       hasOutput = out: builtins.elem out outputs;
-    in
-    {
+    in {
+      repository = getRepository attrs;
+
       # `name` derivation attribute includes cross-compilation cruft,
       # is under assert, and is sanitized.
       # Let's have a clean always accessible version here.
```
</details>

<details><summary>prlist2.diff</summary>

```diff
diff --git a/pkgs/stdenv/generic/check-meta.nix b/pkgs/stdenv/generic/check-meta.nix
index 1cd1ae6dd72e..a28f0c342ecd 100644
--- a/pkgs/stdenv/generic/check-meta.nix
+++ b/pkgs/stdenv/generic/check-meta.nix
@@ -10,7 +10,9 @@ let
     concatMapStrings
     concatMapStringsSep
     concatStrings
+    filter
     findFirst
+    head
     isDerivation
     length
     concatMap
@@ -303,6 +305,10 @@ let
       (listOf str)
       str
     ];
+    repository = union [
+      (listOf str)
+      str
+    ];
     downloadPage = str;
     changelog = union [
       (listOf str)
@@ -440,6 +446,20 @@ let
     # -----
     else { valid = "yes"; });
 
+  # get the default value for the meta.repository field.
+  # the fetchFrom* fetchers set src.meta.homepage
+  getRepository = let
+    getSrcs = attrs:
+      if attrs ? src
+      then
+        [ attrs.src ]
+      else
+        filter (src: src ? meta.homepage) attrs.srcs;
+    getHomePages = map (src: src.meta.homepage);
+    unlist = list:
+      if length list == 1 then head list
+      else list;
+    in attrs: unlist (getHomePages (getSrcs attrs));
 
   # The meta attribute is passed in the resulting attribute set,
   # but it's not part of the actual derivation, i.e., it's not
@@ -452,8 +472,9 @@ let
     let
       outputs = attrs.outputs or [ "out" ];
       hasOutput = out: builtins.elem out outputs;
-    in
-    {
+    in {
+      repository = optionals (attrs ? src.meta.homepage || attrs ? srcs && isList attrs.srcs && any (src: src ? meta.homepage) attrs.srcs) (getRepository attrs);
+
       # `name` derivation attribute includes cross-compilation cruft,
       # is under assert, and is sanitized.
       # Let's have a clean always accessible version here.
```
</details>
<details><summary>prlist3.diff</summary>

```diff
diff --git a/pkgs/stdenv/generic/check-meta.nix b/pkgs/stdenv/generic/check-meta.nix
index 1cd1ae6dd72e..4c1f6ce8e81f 100644
--- a/pkgs/stdenv/generic/check-meta.nix
+++ b/pkgs/stdenv/generic/check-meta.nix
@@ -10,7 +10,9 @@ let
     concatMapStrings
     concatMapStringsSep
     concatStrings
+    filter
     findFirst
+    head
     isDerivation
     length
     concatMap
@@ -303,6 +305,10 @@ let
       (listOf str)
       str
     ];
+    repository = union [
+      (listOf str)
+      str
+    ];
     downloadPage = str;
     changelog = union [
       (listOf str)
@@ -440,7 +446,6 @@ let
     # -----
     else { valid = "yes"; });
 
-
   # The meta attribute is passed in the resulting attribute set,
   # but it's not part of the actual derivation, i.e., it's not
   # passed to the builder and is not a dependency.  But since we
@@ -452,8 +457,17 @@ let
     let
       outputs = attrs.outputs or [ "out" ];
       hasOutput = out: builtins.elem out outputs;
-    in
-    {
+    in {
+      repository = let
+        srcs =
+          if attrs ? src then [ attrs.src ]
+          else filter (src: src ? meta.homepage) attrs.srcs;
+        unlist = list:
+          if length list == 1 then head list
+          else list;
+        hasRep = attrs ? src.meta.homepage || attrs ? srcs && isList attrs.srcs && any (src: src ? meta.homepage) attrs.srcs;
+        in optionals hasRep (unlist (map (src: src.meta.homepage) srcs));
+
       # `name` derivation attribute includes cross-compilation cruft,
       # is under assert, and is sanitized.
       # Let's have a clean always accessible version here.
```
</details>

<details><summary>prlist4.diff</summary>

```diff
diff --git a/pkgs/stdenv/generic/check-meta.nix b/pkgs/stdenv/generic/check-meta.nix
index 1cd1ae6dd72e..7895b1b2f5b9 100644
--- a/pkgs/stdenv/generic/check-meta.nix
+++ b/pkgs/stdenv/generic/check-meta.nix
@@ -10,7 +10,9 @@ let
     concatMapStrings
     concatMapStringsSep
     concatStrings
+    filter
     findFirst
+    head
     isDerivation
     length
     concatMap
@@ -303,6 +305,10 @@ let
       (listOf str)
       str
     ];
+    repository = union [
+      (listOf str)
+      str
+    ];
     downloadPage = str;
     changelog = union [
       (listOf str)
@@ -440,7 +446,6 @@ let
     # -----
     else { valid = "yes"; });
 
-
   # The meta attribute is passed in the resulting attribute set,
   # but it's not part of the actual derivation, i.e., it's not
   # passed to the builder and is not a dependency.  But since we
@@ -452,8 +457,16 @@ let
     let
       outputs = attrs.outputs or [ "out" ];
       hasOutput = out: builtins.elem out outputs;
-    in
-    {
+    in {
+      repository = let
+        srcs =
+          if attrs ? src then [ attrs.src ]
+          else filter (src: src ? meta.homepage) attrs.srcs;
+        unlist = list:
+          if length list == 1 then head list
+          else list;
+        in optionals (attrs ? src.meta.homepage || attrs ? srcs && isList attrs.srcs && any (src: src ? meta.homepage) attrs.srcs) (unlist (map (src: src.meta.homepage) srcs));
+
       # `name` derivation attribute includes cross-compilation cruft,
       # is under assert, and is sanitized.
       # Let's have a clean always accessible version here.
```
</details>

**Comment by nyabinary** [comment]
This needs to be squashed a bit btw.

**Comment by lolbinarycat** [comment]
i will squash once i'm sure everything is working properly.

File: pkgs/stdenv/generic/check-meta.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/stdenv/generic/check-meta.nix:468]: the below diff `prlist11.diff` shows the minimal perf impact as seen in comment https://github.com/NixOS/nixpkgs/pull/300313#issuecomment-2031023142. (by about a factor of 2 to the current change `prlist`) and doesn't nest the conditions. putting the `unlist` function in a let under `repository` increases the geomean of `prlist11` from 0.13 -> 0.14 (not shown)

<details><summary>prlist11.diff</summary>

```diff
diff --git a/pkgs/stdenv/generic/check-meta.nix b/pkgs/stdenv/generic/check-meta.nix
index 1cd1ae6dd72e..f1f486b115a8 100644
--- a/pkgs/stdenv/generic/check-meta.nix
+++ b/pkgs/stdenv/generic/check-meta.nix
@@ -10,7 +10,9 @@ let
     concatMapStrings
     concatMapStringsSep
     concatStrings
+    filter
     findFirst
+    head
     isDerivation
     length
     concatMap
@@ -39,6 +41,11 @@ let
     toPretty
   ;
 
+  unlist = list:
+  if length list == 1
+  then head list
+  else list;
+
   # If we're in hydra, we can dispense with the more verbose error
   # messages and make problems easier to spot.
   inHydra = config.inHydra or false;
@@ -303,6 +310,10 @@ let
       (listOf str)
       str
     ];
+    repository = union [
+      (listOf str)
+      str
+    ];
     downloadPage = str;
     changelog = union [
       (listOf str)
@@ -452,8 +463,14 @@ let
     let
       outputs = attrs.outputs or [ "out" ];
       hasOutput = out: builtins.elem out outputs;
-    in
-    {
+    in {
+      repository =
+        if attrs ? src.meta.homepage
+        then attrs.src.meta.homepage
+        else if attrs ? srcs && isList attrs.srcs
+        then unlist (map (src: src.meta.homepage) (filter (src: src ? meta.homepage) attrs.srcs))
+        else [];
+
       # `name` derivation attribute includes cross-compilation cruft,
       # is under assert, and is sanitized.
       # Let's have a clean always accessible version here.
```
</details>

File: pkgs/applications/graphics/sane/backends/brscan4/default.nix
Comment (lolbinarycat) [reply pkgs/applications/graphics/sane/backends/brscan4/default.nix:93]: > the eval for unsupported platforms shouldn't fail anymore.

i'm not sure what makes you think that? brscan4 still has a src that fails to eval on unsupported platforms..

i'm still going to change this, because i think making src eval on all platforms is a cleaner fix.

Comment (a-n-n-a-l-e-e) [reply pkgs/applications/graphics/sane/backends/brscan4/default.nix:93]: > > the eval for unsupported platforms shouldn't fail anymore.
> 
> i'm not sure what makes you think that? brscan4 still has a src that fails to eval on unsupported platforms..

```
nix-instantiate -A brscan4 --show-trace --argstr system aarch64-linux
```

File: pkgs/applications/graphics/sane/backends/brscan4/default.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/applications/graphics/sane/backends/brscan4/default.nix:25]: fixing these seems out of scope of this PR and not required for meta repository, in my opinion. there are other similar uses in the source tree, eg: `outfox` and there aren't any checks to prevent PRs from adding a src which can not be evaluated on some systems.

i don't see what the downside is of just leaving these cases as is.

Comment (lolbinarycat) [reply pkgs/applications/graphics/sane/backends/brscan4/default.nix:25]: outfox uses `throw`, which can be caught using tryEval.  a throw error doesn't cause ofborg-eval to fail, since there is a tryEval in `release-attrpaths-superset.nix`.

the main user of this field will be nixos-search, which requires `meta` to be fully evaluated and converted into json, although i suppose it's probably fine as long as that is possible on x86_64-linux.

File: pkgs/applications/graphics/sane/backends/brscan4/default.nix
Comment (lolbinarycat) [reply pkgs/applications/graphics/sane/backends/brscan4/default.nix]: hmm, i see, i was actually testing with `nix eval .#pkgsCross.loongarch64-linux.brscan4.meta.repository`, which produces an error.  i suppose i can remove this altogether, hopefully it doesn't cause another hydra build failure down the line.

File: pkgs/development/coq-modules/serapi/default.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/development/coq-modules/serapi/default.nix:75]: this is not required

File: pkgs/servers/photoprism/libtensorflow.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/servers/photoprism/libtensorflow.nix:86]: does not seem required. i am not sure how to reach this attribute through photoprism

**Comment by lolbinarycat** [comment]
@a-n-n-a-l-e-e updated

**Comment by a-n-n-a-l-e-e** [comment]
can you remove the files from the PR -- now the files just have comments which seem outdated. this PR should have 1 file and then squash the commits so we can move forward. thx.

**Comment by lolbinarycat** [comment]
what do you mean by "remove the files from the pr"?

will squashing do that, or do i need to do something else?

File: pkgs/development/coq-modules/serapi/default.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/development/coq-modules/serapi/default.nix:71]: this file can be removed from the chageset

File: pkgs/servers/photoprism/libtensorflow.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/servers/photoprism/libtensorflow.nix:80]: this file too.

**Review by ghost — COMMENTED** [review]
please remove the changes to serapi and photoprism/libsensorflow.nix and then squash the commits. i don't see a reason for these files to be part of the change now.

**Comment by lolbinarycat** [comment]
@a-n-n-a-l-e-e does this look good to you?

**Review by rhendric — CHANGES_REQUESTED** [review]
Please let's get some reviewers on this who didn't work on the previous iteration before merging, to make sure this is a consensus-driven change.

I've been expressing skepticism about this approach in the parent issue and while I don't mind being overruled by consensus, right now it seems like the same people who cowboyed through the first PR in a very short time are pushing forward again without changing much about their approach. I'm not trying to criticize them for their enthusiasm but I'd feel better if some more experienced voices were involved.

**Comment by lolbinarycat** [comment]
> very short time

21 days of active discussion.

> without changing much about their approach

it has been completely rewritten and the semantics have been changed.

> I'd feel better if some more experienced voices were involved

if anyone with relevant experience with stdenv is available in a timely manner, then sure, but i'm not sure that's the case.

**Comment by rhendric** [comment]
> > very short time
> 
> 21 days of active discussion.

\<sigh> @AndersonTorres proposed splitting the PR into two parts along the basis of whether `repository` should be populated from `src` on March 28. @a-n-n-a-l-e-e merged the PR on March 29. That was a very short time to consider and reject that approach.

> > without changing much about their approach
> 
> it has been completely rewritten and the semantics have been changed.

The *approach* of deriving this field from `src` hasn't changed.

> > I'd feel better if some more experienced voices were involved
> 
> if anyone with relevant experience with stdenv is available in a timely manner, then sure, but i'm not sure that's the case.

This isn't how consensus-driven projects should work. Cooperate and build consensus; don't just rush through what you can get away with because you have more free time than the older hands.

**Comment by a-n-n-a-l-e-e** [comment]
> > > very short time
> > 
> > 
> > 21 days of active discussion.
> 
> <sigh> @AndersonTorres proposed splitting the PR into two parts along the basis of whether `repository` should be populated from `src` on March 28. @a-n-n-a-l-e-e merged the PR on March 29. That was a very short time to consider and reject that approach.

this is likely true. i stopped reading @AndersonTorres comments as they were toxic and trollish.

**Comment by lolbinarycat** [comment]
> Cooperate and build consensus

the current PR is an amalgamation of code written by at least three different people, using feedback provided by several others.  if that's not cooperation, i don't know what is.

additionally, i'm not sure how to cooperate with someone when their only feedback is "don't do this".

**Comment by a-n-n-a-l-e-e** [comment]
> @a-n-n-a-l-e-e does this look good to you?

the commit still needs to get squashed. if you're struggling with git one thing i do is create a test branch to experiment on so if i run the wrong git commands i can just toss the branch and start over. eg: `git switch -c test-branch meta.repository` then experiment on `test-branch`.

**Comment by lolbinarycat** [comment]
@a-n-n-a-l-e-e whenever i would squash it would just create a merge conflict, which i would fix on github, but that would just create another commit.

i have 3 backup presquash branches by now

my local nixpkgs clone is extremely messy due to using things like shallow pulls in order to cope with my extremely slow internet.

**Comment by a-n-n-a-l-e-e** [comment]
> @a-n-n-a-l-e-e whenever i would squash it would just create a merge conflict, which i would fix on github, but that would just create another commit.
> 
> i have 3 backup presquash branches by now
> 
> my local nixpkgs clone is extremely messy due to using things like shallow pulls in order to cope with my extremely slow internet.

ok. i fixed it then. you'll need to pull before doing any changes.

File: pkgs/stdenv/generic/check-meta.nix
Comment (AndersonTorres) [inline pkgs/stdenv/generic/check-meta.nix:465]: Not sure, but afaicr format rules RFC requires `in` in a separate line.

File: doc/stdenv/meta.chapter.md
Comment (AndersonTorres) [inline doc/stdenv/meta.chapter.md:52]:
```suggestion
A webpage (or a list of webpages) where the package's source code can be viewed.  `https` links are preferred if available.  Automatically set to a default value if the package uses a `fetchFrom*` fetcher for its `src`. Example: `https://github.com/forthy42/gforth`
```

P.S.: Again, this is not the official upstream source for Gforth. 

And, since the xzdoor incident (that happened on its github mirror), we should not promote mirrors.

**Comment by a-n-n-a-l-e-e** [comment]
@rhendric you have requested changes but have not commented on any part of the PR. This is a low stakes change that has minimal impact on eval performance, https://github.com/NixOS/nixpkgs/pull/300313#issuecomment-2031023142, and does not impact the building of nix packages. additionally, backing out the change, if required, does not require a staging cycle or any rebuilds, so your behavior seems out of place given impact of this change. Please reconsider to help move forward and deescalate this situation or consider stepping down from your role as a committer if you are unable to do so.

**Comment by lolbinarycat** [comment]
so, turns out shallow fetches kinda break everything.  will manually reconstruct the broken branches on top of current master and create a new PR.

**Comment by 7c6f434c** [comment]
@a-n-n-a-l-e-e 1. any committer can dismiss «changes requested» status with a comment, so you might be overstating the level of blocking, and 2. if a change has broken eval and was reverted, its updated version is not low-stakes.

**Pull request closed without merging**
